src/main/java/rx/internal/operators/OperatorObserveOn.java

```diff
@@ -191,41 +188,45 @@ void pollQueue() {
                  */
                 counter = 1;
 
-//                middle:
                 while (!scheduledUnsubscribe.isUnsubscribed()) {
-                    if (failure) {
-                        child.onError(error);
-                        return;
-                    } else {
-                        if (requested == 0 && completed && queue.isEmpty()) {
+                    //make a local copy of the volatile finished
+                    //so that it is only read once in this loop 
+                    boolean _finished = finished;
+                    if (_finished) {
+                        if (error != null) {
+                            //even if there are onNext in the queue we eagerly notify of error
+                            child.onError(error);
+                            return;
+                        } else if (queue.isEmpty()) {
                             child.onCompleted();
                             return;
                         }
-                        if (REQUESTED.getAndDecrement(this) != 0) {
-                            Object o = queue.poll();
-                            if (o == null) {
-                                if (completed) {
-                                    if (failure) {
-                                        child.onError(error);
-                                    } else {
-                                        child.onCompleted();
-                                    }
+                    }
+                    if (REQUESTED.getAndDecrement(this) != 0) {
+                        Object o = queue.poll();
+                        if (o == null) {
+                            // nothing in queue (but be careful, something could be added concurrently right now)
+                            if (_finished) {
```

Comment:
Now this is unnecessary because if queue was empty, it would have been caught a few lines up and might not worth the eagerness as there is only a single atomic update separating the two.